src/Api/Authorization/EndpointConventionBuilderExtensions.cs

```diff
@@ -0,0 +1,77 @@
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Authorization.Infrastructure;
+using Passwordless.Common.Constants;
+using Passwordless.Common.Extensions;
+
+namespace Passwordless.Api.Authorization;
+
+public static class EndpointConventionBuilderExtensions
+{
+    /// <summary>
+    /// Requires a `ManagementKey` header to access the endpoint.
+    /// </summary>
+    /// <param name="builder"></param>
+    /// <typeparam name="TBuilder"></typeparam>
+    /// <returns></returns>
+    public static TBuilder RequireManagementKey<TBuilder>(this TBuilder builder) where TBuilder : IEndpointConventionBuilder
+    {
+        var requirements = new List<IAuthorizationRequirement>
+        {
+            new ClaimsAuthorizationRequirement(CustomClaimTypes.KeyType, new []{ Constants.ManagementKeyType }),
+        };
+
+        var schemes = new List<string> { Constants.ManagementKeyAuthenticationScheme };
+        var pol = new AuthorizationPolicy(requirements, schemes);
+        return builder.RequireAuthorization(pol);
+    }
+
+    /// <summary>
+    /// Requires a `ApiKey` header to access the endpoint.
+    /// </summary>
+    /// <param name="builder"></param>
+    /// <param name="scope">When provided, requires the `ApiKey` to have the specified scope.</param>
+    /// <typeparam name="TBuilder"></typeparam>
+    /// <returns></returns>
+    public static TBuilder RequirePublicKey<TBuilder>(this TBuilder builder, PublicKeyScopes? scope = null) where TBuilder : IEndpointConventionBuilder
+    {
+        var requirements = new List<IAuthorizationRequirement>
+        {
+            new ClaimsAuthorizationRequirement(CustomClaimTypes.AccountName, null),
+            new ClaimsAuthorizationRequirement(CustomClaimTypes.KeyType, new []{ Constants.PublicKeyType }),
+        };
+
+        if (scope.HasValue)
```

Comment:
If a scope was required by an endpoint, add the `ClaimAuthorizationRequirement` to check the claim is present for the authenticated `ApiKey` or `SecretKey`.